pkg/types/field_type.go

```diff
@@ -179,7 +179,7 @@ func SetTypeFlag(flag *uint, flagItem uint, on bool) {
 // InferParamTypeFromDatum is used for plan cache to infer the type of a parameter from its datum.
 func InferParamTypeFromDatum(d *Datum, tp *FieldType) {
 	InferParamTypeFromUnderlyingValue(d.GetValue(), tp)
-	if IsStringKind(d.k) {
+	if IsStringKind(d.k) || d.GetValue() == nil {
```

Comment:
Can you explain why I need to set the charset again here when I have already set it in the `InferParamTypeFromUnderlyingValue` function?